solr/core/src/java/org/apache/solr/core/FSTracerConfigurator.java

```diff
@@ -0,0 +1,124 @@
+package org.apache.solr.core;
+
+import javax.servlet.http.HttpServletRequest;
+import java.lang.invoke.MethodHandles;
+
+import io.jaegertracing.internal.JaegerSpan;
+import io.jaegertracing.internal.JaegerSpanContext;
+import io.jaegertracing.internal.JaegerTracer;
+import io.jaegertracing.internal.reporters.LoggingReporter;
+import io.jaegertracing.internal.reporters.NoopReporter;
+import io.opentracing.Scope;
+import io.opentracing.ScopeManager;
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.opentracing.propagation.Format;
+import org.apache.solr.common.util.NamedList;
+import io.jaegertracing.internal.samplers.ConstSampler;
+import org.apache.solr.util.tracing.GlobalTracer;
+import org.apache.solr.util.tracing.HttpServletCarrier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.MDC;
+
+public class FSTracerConfigurator extends TracerConfigurator  {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  private static final ThreadLocal<Boolean> IS_FORWARDED = ThreadLocal.withInitial(() -> Boolean.FALSE);
+
+  private volatile Tracer tracer;
+  @Override
+  public Tracer getTracer() {
+    return tracer;
+  }
+
+  public void requestStart(HttpServletRequest request) {
+    String reqId = MDC.get(_REQ_ID);
+    if (reqId == null) return;
+    if (log.isInfoEnabled()) {
+      String type = IS_FORWARDED.get() ? "forwarded" : "external";
```

Comment:
using this can we identify top level of request? (top query request vs subrequest for shard)

Reply:
yes